src/NUnitEngine/nunit.engine.core/Services/ExtensionManager.cs

```diff
@@ -363,32 +412,29 @@ private void ProcessAddinsFile(IFile addinsFile, bool fromWildCard)
 
         private void ProcessCandidateAssembly(string filePath, bool fromWildCard)
         {
-            if (WasVisited(filePath, fromWildCard)) 
+            // Did we already process this file?
+            if (_assemblies.ByPath.ContainsKey(filePath))
                 return;
 
-            MarkAsVisited(filePath, fromWildCard);
-
             try
             {
-                var candidate = new ExtensionAssembly(filePath, fromWildCard);
+                var candidateAssembly = new ExtensionAssembly(filePath, fromWildCard);
+                string assemblyName = candidateAssembly.AssemblyName;
 
-                if (!CanLoadTargetFramework(Assembly.GetEntryAssembly(), candidate))
+                // We never add assemblies unless the host can load them
+                if (!CanLoadTargetFramework(Assembly.GetEntryAssembly(), candidateAssembly))
                     return;
-
-                for (var i = 0; i < _assemblies.Count; i++)
+                
+                // Do we already have a copy of the same assembly at a diffrent path?
+                if (_assemblies.ByName.ContainsKey(assemblyName))
                 {
-                    var assembly = _assemblies[i];
+                    if (candidateAssembly.IsBetterVersionOf(_assemblies.ByName[assemblyName]))
+                        _assemblies.ByName[assemblyName] = candidateAssembly;
```

Comment:
Could use `TryGetValue` to prevent need for double lookup.
```csharp
                if (_assemblies.ByName.TryGetValue(assemblyName, out ExtensionAssembly existing))
                {
                    if (candidateAssembly.IsBetterVersionOf(existing))
```

Reply:
Nice. I really ought to wrap that logic in the ExtensionAssemblyTracker class. Later though. :-)